### PR TITLE
Selenium

### PR DESCRIPTION
**Pull request by TravisHunting**
Major reworks, lots of new features. Reworked the original features some and haven't tested them, have only tested with the Selenium functions

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -34,7 +33,9 @@ optional arguments:
   -l, --lowres          Download low resolution images. Omit this argument to download the max
                         quality images
   -d, --disable-wait    Disable the wait between requests (captcha guard)
-
+  -s, --selenium        Scrape image links using Selenium and a headless browser
+  -sd, --selenium-display
```
Comment (AustinHunting) [inline README.md:37]: not loving the `-s` and `-sd` separation commands, I vote we stash the less used one as a double slash:
```
parser.add_argument('--display', help='Use Selenium in display mode', action='store_true')
```

File: comicScraper.py [hunk 2]
```diff
@@ -272,13 +424,20 @@ def main(fullComicDownload, singleIssueDownload, title, lowres, disableWait):
     parser.add_argument('-c', '--complete', help='Download the entire comic into one folder. Omit this argument to download each issue into its own folder', action='store_true')
     parser.add_argument('-l', '--lowres', help='Download low resolution images. Omit this argument to download the max quality images', action='store_true')
     parser.add_argument('-d', '--disable-wait', help='Disable the wait between requests (captcha guard)', action='store_true')
+    parser.add_argument('-s', '--selenium', help='Scrape image links using Selenium and a headless browser', action='store_true')
+    parser.add_argument('-sd', '--selenium-display', help='Use Selenium in display mode', action='store_true')
```
Comment (AustinHunting) [inline comicScraper.py:428]: not loving the `-s` and `-sd` separation commands, I vote we stash the less used one as a double slash:
```
parser.add_argument('--display', help='Use Selenium in display mode', action='store_true')
```

**Review by AustinHunting — COMMENTED** [review]
running this with `./comicScraper.py -sd https://readcomiconline.li/Comic/Fables` gets me:
```
...
...
...
...
====== WebDriver manager ======
Current google-chrome version is 97.0.4692
Get LATEST chromedriver version for 97.0.4692 google-chrome
Driver [/home/aus/.wdm/drivers/chromedriver/linux64/97.0.4692.71/chromedriver] found in cache

Traceback (most recent call last):
  File "./comicScraper.py", line 488, in <module>
    main(downloadFull, singleIssue, comicTitle, lowres, disableWait, startURL, useSelenium, seleniumDisplay)
  File "./comicScraper.py", line 385, in main
    imageLinks, issueImageDict = downloadAllWithSelenium(fullComicDownload, startURL, issueLinks, title, singleIssueDownload, disableWait, seleniumDisplay)
  File "./comicScraper.py", line 312, in downloadAllWithSelenium
    options.add_extension('AdblockPlusModified.crx')
  File "/home/aus/.local/lib/python3.8/site-packages/selenium/webdriver/chromium/options.py", line 101, in add_extension
    raise IOError("Path to the extension doesn't exist")
OSError: Path to the extension doesn't exist
```

File: comicScraper.py [hunk 3]
```diff
@@ -194,26 +239,96 @@ def saveImageFromUrl(url, numberOfImages, issueName=""):
...
+    counter=random.randint(10,20)
+    print(f"Sleeping for {counter} seconds")
+    time.sleep(counter)
+
+def downloadIssueWithSelenium(fullComicDownload, driver, service, issue, imageLinks, issueImageDict, startURL, title, singleIssueDownload, disableWait, seleniumDisplay):
+    issueName = getIssueName(issue, startURL)
+    # I *think* that driver.get causes it to wait until the entire page finish loading
+    # It *appears* to load all the images into the browser before continuing
+    print(f"Selenium accessing {issue}")
+    print("Please wait...")
```
Comment (AustinHunting) [inline comicScraper.py:253]: I feel like we should hide these behind a `log.debug` or `log.info` option. Not sure if we really need to inform the end user of this.

File: comicScraper.py [hunk 4]
```diff
@@ -162,11 +166,52 @@ def extractImageUrlFromText(text, lowres):
         print("extractImageUrlFromText output ", output)
     return output
 
+def displayDownloadInfo(path):
+    list_of_files = list(filter( lambda x: os.path.isfile(os.path.join(path, x)), os.listdir(path)))
+    files_with_size = [ (file_name, os.stat(os.path.join(path, file_name)).st_size) for file_name in list_of_files  ]
+
+    print(f"Downloaded {len(list_of_files)} files to {path}")
```
Comment (AustinHunting) [inline comicScraper.py:173]: path can probably be hidden under `log.debug` or if the user gives us a file path with a with a flag

File: comicScraper.py [hunk 5]
```diff
@@ -162,11 +166,52 @@ def extractImageUrlFromText(text, lowres):
...
+        sizeDisplay = round(((totalSize/1000)/len(list_of_files)),3)
+        sizeType = "kb"
+
+    print(f"Average file size: {sizeDisplay}{sizeType}")
+
+    if len(smallFiles) > 0:
+        print(f"The following {len(smallFiles)} files are smaller than 10kb")
+        print(smallFiles)
+    else:
+        print("All files are larger than 10kb - good to go!")
```
Comment (AustinHunting) [inline comicScraper.py:205]: Absolutely loving the checking here, seems like a great way to verify that the right files are pulled plus the log message is helpful for seeing how the download is progressing

File: comicScraper.py (hunk 4 above)
Comment (AustinHunting) [reply comicScraper.py:173]: granted I really like the `{len(list_of_files)}` portion 
maybe:
```
if log.status == debug or arguments.folder:
    print(f"Downloaded {len(list_of_files)} files to {path}")
else:
    print(f"Downloaded {len(list_of_files)}")
```
(double check my args and spacing, all in the web  editor here)

File: comicScraper.py [hunk 6]
```diff
@@ -162,11 +166,52 @@ def extractImageUrlFromText(text, lowres):
...
+
 def saveImagesFromImageLinks(imageLinks, numberOfImages, issueName=""):
+    global COUNTER
+    initial = COUNTER
```
Comment (AustinHunting) [inline comicScraper.py:212]: Just a note in general, we really need to fix this as passed args
@AustinHunting

File: comicScraper.py [hunk 7]
```diff
@@ -307,7 +466,24 @@ def main(fullComicDownload, singleIssueDownload, title, lowres, disableWait):
...
+    disableWait = False
+    if arguments.disable_wait == True:
+        print("Argument -d detected. Disabling wait between requests")
+        print("This may cause CAPTCHAs to appear more often.")
+        disableWait = True
+
     print(f"Starting to scrape {comicTitle} from {startURL}")
 
-    main(downloadFull, singleIssue, comicTitle, lowres, arguments.disable_wait)
+    main(downloadFull, singleIssue, comicTitle, lowres, disableWait, startURL, useSelenium, seleniumDisplay)
```
Comment (AustinHunting) [inline comicScraper.py:501]: We should probably turn this into a dict/list before passing it around, a massive list of args seems clunky
@AustinHunting

File: comicScraper.py [hunk 8]
```diff
@@ -307,7 +466,24 @@ def main(fullComicDownload, singleIssueDownload, title, lowres, disableWait):
     else:
         print("Downloading max quality images")
 
+    useSelenium = False
+    if arguments.selenium == True:
+        print("Argument -s detected. Using Selenium to scrape the page(s)")
+        useSelenium = True
+    
+    seleniumDisplay = False
+    if arguments.selenium_display == True:
```
Comment (AustinHunting) [inline comicScraper.py:488]: need to add a file check for `AdblockPlusModified.crx` and if not found kick back to just `-s` mode. 

I was running the script in a container and got errored out on this

File: comicScraper.py (hunk 7 above)
Comment (TravisHunting) [reply comicScraper.py:502]: I'm on board

File: comicScraper.py [hunk 9]
```diff
@@ -272,13 +424,20 @@ def main(fullComicDownload, singleIssueDownload, title, lowres, disableWait):
...
+    parser.add_argument('-sd', '--selenium-display', help='Use Selenium in display mode', action='store_true')
 
     # ensure that no args is a help call
     if len(sys.argv)==1:
         parser.print_help(sys.stderr)
         sys.exit(1)
     arguments = parser.parse_args()
 
+    if arguments.selenium == True and arguments.selenium_display == True:
+        print("Please provide only -s or -sd, not both")
```
[the remark this answers is not in the document]
Comment (TravisHunting) [reply comicScraper.py:451]: I'm totally happy for these to be reworked in whatever way you think is best. Should be quick

File: comicScraper.py (hunk 2 above)
Comment (TravisHunting) [reply comicScraper.py:442]: Fine with me, not sure exactly what you mean but definitely go for it

**Review by AustinHunting — COMMENTED** [review]
General note, we need to figure out if we're using the default logging package or loguru. the `-v` flag should be reserved for setting logging from info to debug

**Review by AustinHunting — COMMENTED** [review]
Something to note, the end file are not pre-padded with zeros. `issue-1.cbz` vs `issue-87.cbz`

I realize this might be a trick now that we don't have a global image/issue counter. I propose we do a cleanup at the very end when we remove the `.tmp` file that renames the files, either in raw python or by calling the bash script in the shell directory.

@AustinHunting clean up the bash script for inputs

**Review by AustinHunting — APPROVED** [review]
I don't see anything else of note so I'm going to mark this as approved so we can merge it and branch for each little issue. All in all great job, this was an awesome pr and I'm loving the progress you're making. the selenium part, prime. the new terminal outputs, prime, the file size and count checking, prime. extremely good job on this

**Pull request merged**